pkg/machinehealthcheck/machinehealthcheck.go

```diff
@@ -68,57 +49,104 @@ var _ = Describe("[Feature:MachineHealthCheck] MachineHealthCheck", func() {
 	})
 
 	AfterEach(func() {
+		By("Deleting the MachineHealthCheck resource")
+		Expect(client.Delete(context.Background(), machinehealthcheck)).ToNot(HaveOccurred())
+
 		By("Deleting the new MachineSet")
-		err := client.Delete(context.Background(), machineSet)
-		Expect(err).ToNot(HaveOccurred())
+		Expect(client.Delete(context.Background(), machineSet)).ToNot(HaveOccurred())
 
 		framework.WaitForMachineSetDelete(client, machineSet)
 	})
 
 	It("should remediate unhealthy nodes", func() {
-		By("Setting conditions on a Node")
-
 		selector := machineSet.Spec.Selector
 		machines, err := framework.GetMachines(client, &selector)
 		Expect(err).ToNot(HaveOccurred())
 		Expect(machines).ToNot(BeEmpty())
 
-		machine := machines[0]
-		node, err := framework.GetNodeForMachine(client, machine)
-		Expect(err).ToNot(HaveOccurred())
-
-		err = framework.AddNodeCondition(client, node, nodeCondition)
-		Expect(err).ToNot(HaveOccurred())
+		By("Setting unhealthy conditions on machine nodes, but not exceding maxUnhealthy threshold")
+		unhealthyMachines, healthyMachines := machines[:maxUnhealthy-1], machines[maxUnhealthy-1:]
+		for _, machine := range unhealthyMachines {
+			node, err := framework.GetNodeForMachine(client, machine)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(framework.AddNodeCondition(client, node, nodeCondition)).ToNot(HaveOccurred())
+		}
 
 		By("Creating a MachineHealthCheck resource")
 		mhcParams := framework.MachineHealthCheckParams{
-			Name:   machineSetParams.Name,
-			Labels: machineSetParams.Labels,
+			Name:   machineSet.Name,
+			Labels: machineSet.Labels,
 			Conditions: []mapiv1beta1.UnhealthyCondition{
 				{
 					Type:    E2EConditionType,
 					Status:  corev1.ConditionTrue,
 					Timeout: "1s",
 				},
 			},
+			MaxUnhealthy: &maxUnhealthy,
 		}
 
-		mhc, err := framework.CreateMHC(client, mhcParams)
+		machinehealthcheck, err = framework.CreateMHC(client, mhcParams)
 		Expect(err).ToNot(HaveOccurred())
-		Expect(mhc).ToNot(BeNil())
+		Expect(machinehealthcheck).ToNot(BeNil())
+
+		By("Waiting for each unhealthy machine to be deleted")
+		framework.WaitForMachinesDeleted(client, unhealthyMachines...)
+
+		By("Waiting for MachineDeleted event from MachineHealthCheck for each unhealthy machine")
+		for _, machine := range unhealthyMachines {
+			Expect(framework.WaitForEvent(client, "Machine", machine.Name, "MachineDeleted")).ToNot(HaveOccurred())
+		}
```

Comment:
Should these be the other way around, since the events will happen before the machines are deleted right?

Reply:
This happens roughly at the same time, and the events are persisted for some time, so I left it there. Also, if the machines are not deleted, we'd rather see the timeout on machine deletion, then on some event search in my opinion. It is better for debugging imo

Reply:
That's a reasonable argument, sounds good to me :)